kvbc/src/kvbc_adapter/blocks_deleter_adapter.cpp

```diff
@@ -0,0 +1,69 @@
+// Concord
+//
+// Copyright (c) 2022 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "kvbc_adapter/blocks_deleter_adapter.hpp"
+#include "assertUtils.hpp"
+#include "IntervalMappingResourceManager.hpp"
+
+using concord::performance::ISystemResourceEntity;
+
+namespace concord::kvbc::adapter {
+
+BlocksDeleterAdapter::BlocksDeleterAdapter(std::shared_ptr<concord::kvbc::categorization::KeyValueBlockchain> &kvbc,
+                                           const std::optional<aux::AdapterAuxTypes> &aux_types)
+    : kvbc_{kvbc} {
+  if (aux_types.has_value()) {
+    replica_resources_.reset(&(aux_types->resource_entity_));
+  }
+  ConcordAssertEQ(!kvbc_, false);
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// IBlocksDeleter implementation
+void BlocksDeleterAdapter::deleteGenesisBlock() {
+  const auto genesisBlock = kvbc_->getGenesisBlockId();
+  if (genesisBlock == 0) {
+    throw std::logic_error{"Cannot delete the genesis block from an empty blockchain"};
+  }
+  kvbc_->deleteBlock(genesisBlock);
+}
+BlockId BlocksDeleterAdapter::deleteBlocksUntil(BlockId until) {
+  auto mes_prun_util(replica_resources_ ? std::make_unique<ISystemResourceEntity::scopedDurMeasurment>(
+                                              *replica_resources_, ISystemResourceEntity::type::pruning_utilization)
+                                        : nullptr);
+  const auto genesisBlock = kvbc_->getGenesisBlockId();
+  if (genesisBlock == 0) {
+    throw std::logic_error{"Cannot delete a block range from an empty blockchain"};
+  } else if (until <= genesisBlock) {
+    throw std::invalid_argument{"Invalid 'until' value passed to deleteBlocksUntil()"};
+  }
+
+  const auto lastReachableBlock = kvbc_->getLastReachableBlockId();
+  const auto lastDeletedBlock = std::min(lastReachableBlock, until - 1);
+  const auto start = std::chrono::steady_clock::now();
+  for (auto i = genesisBlock; i <= lastDeletedBlock; ++i) {
+    auto mes_prun_avg(replica_resources_ ? std::make_unique<ISystemResourceEntity::scopedDurMeasurment>(
```

Comment:
In the original code, the scopedDurMeasurment is a local object, 
I think that transforming it to. pointer is expensive as it requires allocation per call.
can the replica_resources_ be a reference? 

as far as I remember the ```replica_resources_``` should always exist on the replica. 
The utilization measurement can be done from the replica and not in the adapter.